### PR TITLE
Use DashMap as the hash map in some places.

### PR DESCRIPTION
**Pull request by grao1991**
### Description
Observed lock contention at these places. Let's start with Dashmap.
We might be able to make additional optimization in a different way (e.g. cache hash, shard manually, etc.) in the future.

### CODE REVIEW

**Review by msmouse — APPROVED** [review]
Some numbers in the summary to give people an impression how much this is helping (and under what circumstances) would be nice.

**Comment by sitalkedia** [comment]
@grao1991 - Looking at the land-blocking TPS, I think this is actually introducing a regression. I recall @gelash also tried similar change in past and we observed regression in land-blocking Forge. Can we compare the get state value counter for storage with and without Dashmap? 

Also, IIUC, the caching in `cached_state_view` is useless today, as the `MVHashmap` already caches storage read, so I would be surprised if any improvement in `cached_state_view` is actually making any difference.

**Comment by grao1991** [comment]
> Some numbers in the summary to give people an impression how much this is helping (and under what circumstances) would be nice.

For the cached state view one, I've observed significant overhead here (even for cache hit) in storage only benchmark (with fake executor simulate reads). Before this change, increasing the number of threads will significantly increase the latency here, from 8 threads -> <10us to 32 threads -> ~60us. With this change we can still get ~10us for cache hit when using 32 threads.

**Review by gelash — COMMENTED** [review]
Can we indeed make sure there is no regression?
If there is a regression, there is one place where we collect with dashmap where things could have gotten slower. Otherwise, if the lock is indeed contended, unclear why it would be significantly worse. 

regarding usefulness though, @sitalkedia in case some item keeps getting read, but not written to, MVHashMap does not cache anything, so the storage cache will be important. Also, isn't `cached_state_view` useful between blocks ( or is it getting flushed too, I don't remember)?

**Comment by gelash** [comment]
@grao1991 storing the hash inside statekey would be awesome if you can get to it, I think that should really help

**Comment by grao1991** [comment]
> 



> Can we indeed make sure there is no regression? If there is a regression, there is one place where we collect with dashmap where things could have gotten slower. Otherwise, if the lock is indeed contended, unclear why it would be significantly worse.
> 
> regarding usefulness though, @sitalkedia in case some item keeps getting read, but not written to, MVHashMap does not cache anything, so the storage cache will be important. Also, isn't `cached_state_view` useful between blocks ( or is it getting flushed too, I don't remember)?

I think it totally depends on how many threads we have. On forge only 8 threads, so the benefits are very minimal comparing to the overhead.

**Comment by zekun000** [comment]
i don’t think the regression comes from this pr, i saw 5k number yesterday from another pr

**Comment by zekun000** [comment]
https://github.com/aptos-labs/aptos-core/pull/6273#issuecomment-1401226431

**Comment by sitalkedia** [comment]
>regarding usefulness though, @sitalkedia in case some item keeps getting read, but not written to, MVHashMap does not cache anything, so the storage cache will be important. Also, isn't cached_state_view useful between blocks ( or is it getting flushed too, I don't remember)?

That's interesting, that we don't cache reads in `MVHashMap`. I guess in that case the storage cache is helpful. It is not cached across blocks BTW.

File: storage/storage-interface/src/cached_state_view.rs [hunk 1]
```diff
@@ -135,7 +136,7 @@ impl CachedStateView {
     pub fn into_state_cache(self) -> StateCache {
         StateCache {
             frozen_base: self.speculative_state,
-            state_cache: self.state_cache.into_inner(),
+            state_cache: self.state_cache.into_iter().collect(),
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/storage-interface/src/cached_state_view.rs:139]: Let me double check this.

[the remark this answers is not in the document]
Comment (grao1991) [reply storage/storage-interface/src/cached_state_view.rs]: This did cost some overhead, not huge but not small enough to ignore (~20ms on a 10K block)
Changed to dashmap as well for now. The overhead later at the caller side should be less than this number.

**Comment by grao1991** [comment]
> @grao1991 storing the hash inside statekey would be awesome if you can get to it, I think that should really help

Let me figure out how to do this in an elegant way later.

**Review by zekun000 — APPROVED** [review]

**Pull request merged**